tests/safety/test_chrysler.py

```diff
@@ -121,5 +123,125 @@ def _speed_msg(self, speed):
     return self.packer.make_can_msg_panda("ESP_8", 0, values)
 
 
+class ChryslerLongitudinalBase(common.PandaSafetyTestBase):
+  # pylint: disable=no-member,abstract-method
+
+  DAS_BUS = 0
+
+  MIN_ENGINE_TORQUE = -500
+  MAX_ENGINE_TORQUE = 500
+  MIN_POSSIBLE_ENGINE_TORQUE = -500
+  MAX_POSSIBLE_ENGINE_TORQUE = 1548
+
+  MIN_ACCEL = -3.5
+  MAX_ACCEL = 2.0
+  MIN_POSSIBLE_ACCEL = -16.0
+  MAX_POSSIBLE_ACCEL = 4.0
+
+  @classmethod
+  def setUpClass(cls):
+    if cls.__name__ == "ChryslerLongitudinalBase":
+      cls.safety = None
+      raise unittest.SkipTest
+
+  # override these tests from PandaCarSafetyTest, chrysler longitudinal uses button enable
+  def test_disable_control_allowed_from_cruise(self):
+    pass
+
+  def test_enable_control_allowed_from_cruise(self):
+    pass
+
+  def test_sampling_cruise_buttons(self):
+    pass
+
+  def test_cruise_engaged_prev(self):
+    pass
+
+  def test_button_sends(self):
+    pass
+
+  def _pcm_status_msg(self, enable):
+    raise Exception
+
+  def _send_torque_accel_msg(self, enable, torque_active, torque, accel_active, accel):
+    values = {
+      "ACC_AVAILABLE": 1,
+      "ACC_ACTIVE": enable,
+      "ENGINE_TORQUE_REQUEST": torque,
+      "ENGINE_TORQUE_REQUEST_MAX": torque_active,
+      "ACC_DECEL": accel,
+      "ACC_DECEL_REQ": accel_active,
+    }
+    return self.packer.make_can_msg_panda("DAS_3", self.DAS_BUS, values)
+
+  def _send_accel_msg(self, accel):
+    return self._send_torque_accel_msg(1, 0, 0, 1, accel)
+
+  def _send_torque_msg(self, torque):
+    return self._send_torque_accel_msg(1, 1, torque, 0, 0)
+
+  def test_accel_torque_safety_check(self):
+    # TODO: test sending torque and accel at the same time? (prevent gas and brakes at the same time?)
+    self._generic_limit_safety_check(self._send_accel_msg,
+                                     self.MIN_ACCEL, self.MAX_ACCEL,
+                                     self.MIN_POSSIBLE_ACCEL, self.MAX_POSSIBLE_ACCEL,
+                                     test_delta=0.1, inactive_value=None)
+    self._generic_limit_safety_check(self._send_torque_msg,
+                                     self.MIN_ENGINE_TORQUE, self.MAX_ENGINE_TORQUE,
+                                     self.MIN_POSSIBLE_ENGINE_TORQUE, self.MAX_POSSIBLE_ENGINE_TORQUE,
+                                     test_delta=10, inactive_value=None)
+
+  def test_buttons(self):
+    button_combinations = set(itertools.permutations([0,0,0,0,1,1,1,1], 4))
```

Comment:
this is an inefficient way to generate all permutations of possible button combinations, is there something better?

Reply:
itertools.chain?